File: cpp/se/ProtocolExtensionCommand.cpp
#include "common/ErrorPkg.hpp"
#include "se/CLTRID.hpp"
#include "se/IllegalArgException.hpp"
#include "se/ProtocolExtensionCommand.hpp"
#include "xml/XMLHelper.hpp"

ProtocolExtensionCommand::ProtocolExtensionCommand(
        const CommandType* commandType,
        const ObjectType* objectType,
        const std::string& ident,
        const Extension& ext) : SendSE(),
                                cmdType(commandType),
                                objType(objectType),
                                extension(ext)
{
    if (commandType == NULL)
    {
        throw IllegalArgException(
                ErrorPkg::getMessage("se.command.type.missing"));
    }
    cmdType = commandType;

    command = xmlWriter->appendChild(
            xmlWriter->appendChild(
                xmlWriter->getRoot(),
                "extension"),
            "command",
            extension.getURI());

    command->setAttribute(XStr("xsi:schemaLocation").str(),
            XStr(extension.getSchemaLocation()).str());

    cmdElement = xmlWriter->appendChild(command, cmdType->getCommandName());

    objElement = xmlWriter->appendChild(cmdElement,
            commandType->getCommandName(),
            objectType->getURI());
    objElement->setAttribute(
            XStr("xsi:schemaLocation").str(),
            XStr(objType->getSchemaLocation()).str());

    DOMElement* element = xmlWriter->appendChild(
            objElement, objectType->getIdentType());
    XMLHelper::setTextContent(element, ident);
}

std::string ProtocolExtensionCommand::toXMLImpl()
{
    XMLHelper::setTextContent(
        xmlWriter->appendChild(command, "clTRID"),
        CLTRID::nextVal());

    return xmlWriter->toXML();
}

File: src/main/java/com/ausregistry/jtoolkit2/se/ProtocolExtensionCommand.java
package com.ausregistry.jtoolkit2.se;

/**
 * Base class for protocol extension commands. Instances of this class are
 * responsible for building the part of the XML DOM tree common to all protocol
 * extension commands.
 */
public abstract class ProtocolExtensionCommand extends ObjectCommand {

    private static final long serialVersionUID = 4842063175523680995L;

    public ProtocolExtensionCommand(CommandType cmdType, ObjectType objType) {
        super(cmdType, objType);
    }

    public ProtocolExtensionCommand(CommandType cmdType, ObjectType objType,
            String ident) {
        super(cmdType, objType, ident);
    }

    public ProtocolExtensionCommand(CommandType cmdType, ObjectType objType,
            String ident, String attrName, String attrValue) {
        super(cmdType, objType, ident, attrName, attrValue);
    }

    @Override
    protected void initCmdElement() {
        command = xmlWriter.appendChild(xmlWriter.appendChild(
                xmlWriter.getRoot(), "extension"), "command",
                getExtension().getURI());

        command.setAttribute("xsi:schemaLocation",
                getExtension().getSchemaLocation());
    }

    protected abstract Extension getExtension();
}
